File: 2023/ruby/day3.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

lines = File.readlines('inputs/3').map(&:chomp)

# rubocop:disable Metrics
def get_numbers(symbol_pos, line_pos, lines)
  start = line_pos
  finish = line_pos
  start = line_pos - 1 if line_pos.positive?
  finish = line_pos + 1 if line_pos < lines.length - 1
  numbers = []

  lines[start..finish].each do |line|
    number = String.new
    number_start = 0
    number_finish = 0
    line.split('').each_with_index do |char, char_index|
      if char =~ /[0-9]/
        number_start = char_index if number.empty?
        number_finish = char_index
        number.concat char
      elsif !number.empty? && (((number_start - symbol_pos).abs <= 1) || ((number_finish - symbol_pos).abs <= 1))
        numbers.push number.to_i
        number.clear
      else
        number.clear
      end
    end
    if !number.empty? && (((number_start - symbol_pos).abs <= 1) || ((number_finish - symbol_pos).abs <= 1))
      numbers.push number.to_i
      number.clear
    end
  end

  numbers
end
# rubocop:enable Metrics

def part1(lines)
  sum = 0
  lines.each_with_index do |line, lines_index|
    line.split('').each_with_index do |char, char_index|
      get_numbers(char_index, lines_index, lines).each { |n| sum += n } unless char =~ /[0-9]|[.]/
    end
  end
  sum
end

def part2(lines)
  sum = 0
  lines.each_with_index do |line, lines_index|
    line.split('').each_with_index do |char, char_index|
      if char =~ /[*]/
        numbers = get_numbers(char_index, lines_index, lines)
        sum += numbers[0] * numbers[1] if numbers.length == 2
      end
    end
  end
  sum
end

puts part1(lines)
puts part2(lines)

File: 2023/rust/src/day3.rs
pub fn main(part: usize) {
    let input = std::fs::read_to_string("inputs/3").unwrap();

    match part {
        0 => part1(input),
        1 => part2(input),
        _ => panic!(),
    }
}

pub fn part2(input: String) {
    let mut sum = 0;
    for (line_index, line) in input.lines().enumerate() {
        for (char_index, character) in line.chars().enumerate() {
            if !character.is_ascii_digit() && character == '*' {
                let gears = get_number(input.clone(), line_index, char_index);
                if gears.len() == 2 {
                    sum += gears[0] * gears[1]
                }
            }
        }
    }
    println!("{sum}");
}

pub fn part1(input: String) {
    let mut sum = 0;
    for (line_index, line) in input.lines().enumerate() {
        for (char_index, character) in line.chars().enumerate() {
            if !character.is_ascii_digit() && character != '.' {
                for number in get_number(input.clone(), line_index, char_index) {
                    sum += number;
                }
            }
        }
    }
    println!("{sum}");
}

fn get_number(input: String, line_index: usize, char_index: usize) -> Vec<u64> {
    let mut lines = Vec::new();
    let mut i = 0;
    if line_index != 0 {
        i = line_index - 1
    }
    for n in i..line_index + 2 {
        let line = input.lines().nth(n);
        if let Some(i) = line {
            lines.push(i);
        }
    }
    let mut numbers: Vec<u64> = Vec::new();

    for line in lines {
        let mut number_start: usize = 0;
        let mut number_end: usize = 0;
        let mut number = String::new();

        for (index, character) in line.chars().enumerate() {
            if character.is_ascii_digit() {
                if number.is_empty() {
                    number_start = index;
                    number_end = index;
                } else {
                    number_end = index;
                }
                number.push(character);
            } else if !character.is_ascii_digit()
                && !number.is_empty()
                && (number_start.abs_diff(char_index) <= 1 || number_end.abs_diff(char_index) <= 1)
            {
                numbers.push(number.parse().unwrap());
                number.clear()
            } else {
                number.clear()
            }
        }
        if !number.is_empty()
            && (number_start.abs_diff(char_index) <= 1 || number_end.abs_diff(char_index) <= 1)
        {
            numbers.push(number.parse().unwrap());
            number.clear();
        }
    }
    numbers
}
